### DISCUSSION TITLE
Merge skeletons from different models

### DISCUSSION THREAD

**Discussion by rfkova** [talmolab/sleap#1316] [Help!] [answered]
Hello,

I have found that the objects I'm modeling using sleap are best detected with separate models. The goal is to create pose estimations with two models on the same video, then merge the two prediction outputs (with different skeletons) so I can submit the merged predictions to SIMBA for behavioral analysis using an amalgam. For reference the objects I'm modelling are a dam and a nest (which changes shape) so these are single instance models.

Many many thanks!
Ryan

**Comment by roomrys** [comment]
Hi @rfkova,

I'm not sure I caught your question, but I'll give some info on merging to see if that helps - let me know if you need more info!

When merging projects, SLEAP will also attempt to resolve skeletons by adding nodes/edges that do not exist in the base project (to merge into). Nodes that have the same name in each differing skeletons/projects will be taken as the same node - this will result in a conflict where the user must decide which node to keep from which project. Thus, it would be best to ensure the nodes in your skeletons have unique node names. After merging, you can add any edges that might be missing or change the configuration of edges.

Thanks,
Liezl

**Comment by rfkova** [comment]
@roomrys,

Liezl! idk if you remember but we had a legendary troubleshooting episode quite some time ago now. I still think back on that whole experience positively despite how frustrating it was, so pumped to see you still working on this (hope you are still enjoying the work)!

Sorry for not stating a question directly - I was asking how to merge two files in the fashion I described. My problem is that the Merge into Project option (under File) gives me a pop-up after file selection and none of the options merge both labels into one file. The nodes are created, but I have to choose labels from one video or the other or to discard all overlapping labels (Use New..., Use Base..., Discard all...). Is there a way to keep both? The node names are fully unique between files I want to merge. 

It sounds like you thought it would do what I wanted... Do you think there's something wrong with my approach? Is there anything I can try? 

I sent a screenshot of the options I have after the two files load and all of these options discard tons of predictions according to how the options are described.

![image](https://github.com/talmolab/sleap/assets/64923232/22b88459-6a11-4e39-91ac-2b2068873ba7)

Best,
Ryan

**Reply by roomrys** [reply]
Hi @rfkova,

Yes, I thought that was you! Since that session, I've coined your reference to a "monster skeleton".

I tested this over here and although the new nodes are added to the skeleton, we just throw away the locations from the nodes! We should include another option to keep location info on uniquely named nodes (which I am sure others will also find very useful). I'll work on adding this.

One implementation difficulty I see is merging multiple instances together... thus, this might only be a feature available to single animal projects.

Thanks,
Liezl

**Reply by rfkova** [reply]
@roomrys,

Amazing, thank you! What is the best way to keep apprised of when such a feature may become available? I would be happy for a non-GUI implementation as well if there is some arrangement of code I could leverage, just sayin'. A pie in the sky solution would be to select categories of files to combine from different folders or something, but if the code exists for a single pair of label files it will be pretty easy to target and loop pairs - not asking for that on your end but just a suggestion in case that seems like something that could be more broadly useful. 

re: "monster skeleton" or my other favorite the "Cronenberg skeleton", I love it! Honestly if my only contribution to the world is that this somehow leaks out into broader pose-estimation circles (as broadly as that goes (◔_◔) ), I feel my efforts have been well worth it.

Cheers,
Ryan

**Reply by roomrys** [reply]
Hi @rfkova,

We are working on getting a release out end of this week, so this feature likely will not make the cut for this release... To keep you rolling, I could make a "not for production" script to test out (on copies/back-ups of your data - of course). I will post that here when it is ready (aiming for tomorrow or Friday!).

Thanks,
Liezl

**Reply by rfkova** [reply]
Would love a script to test out! Even if the data is not ultimately usable I could really use this to set up a workflow we are working on.

Even if it doesn't happen (or happen soon) your whole team are GD heroes and like 70% of that impression is from your support.

Best,
Ryan

**Reply by roomrys** [reply]
Hi @rfkova,

I haven't forgotten, but am just a bit behind! Doing some work on the script now.

Thanks,
Liezl

**Reply by rfkova** [reply]
Hi @roomrys,

Thank you so much for working on this 😁

If I can mess around with a script by Friday and get an approximation of well integrated predictions originating from different skeletons that would be a resounding success. If it happens before the end of time, I will still be very grateful haha, so no worries, I didn't think for a moment I had been forgotten!

**Reply by rfkova** [reply]
Hi @roomrys!

Wanted to check in on a rough timeframe. Again, I'm more than pleased to be getting any kind of help on something like this, so no pressure for speed per se, I only want to plan ahead.

Thanks again!
Ryan

**Reply by rfkova** [reply]
Hi @roomrys,

Just pinging this discussion in case it has fallen off the radar.

Best,
Ryan

**Reply by roomrys** [reply]
Back on the radar! I know you have been extremely patient, and I am very grateful for that.

**Reply by rfkova** [reply]
Huzzah! I see how much you've been doing in the change log, so no shade haha. Thanks again for all your work!

**Reply by rfkova** [reply]
@roomrys, 

I know nothing is certain, but do you think some version of a script might be available relatively soon?

I have a student doing some work on the project for the next few weeks so I want to plan.

Would some data to merge be helpful? Anything else I can do?

**Reply by roomrys** [reply]
Hi @rfkova,

Definitely not this week as I will be out the rest of the week. Next week is more likely. I have some data that I was testing merging with before the dependency issues pulled me away. I found that the merging code was a bit convoluted (at least to me) and some changed need to be made prior to getting a script for you that would work as expected.

Liezl

**Reply by talmo** [reply, marked as answer]
Hi @rfkova,

Since Liezl is out and since I'm procrastinating on a grant proposal, I thought I'd take a stab at this.

As Liezl mentioned, the issue is that it's not clear how to handle merge conflicts within a single frame. There are multiple options depending on your use case, for example, whether any nodes overlap, whether you're trying to merge predictions or user labels, and where the data is coming from.

Merging new nodes onto an existing instance will be really hard. First you need to figure out which instance corresponds to which existing instance in the same frame, which is only possible if there are at least some nodes in common between the two -- and even then, it's not clear what to do if they're off by a little (e.g., two predicted instances with different guesses about where the common nodes are).

Assuming we have no duplicate instances or completely non-overlapping skeletons, then it's possible to do this by just merging in every instance from the new labels. I can see this being useful if you have a model of the animals and a separate model of objects (or adults and pups) where the skeletons don't overlap inherently. I still don't see the value of merging the two labels though.

If you _do_ merge them this way, then you might end up with some of your instances missing the new nodes, and some missing the old ones, which is equivalent to inconsistent labeling and will seriously confuse the neural networks if you use this to train.

Anyway, here's a notebook that does this with all these caveats: https://colab.research.google.com/drive/1OmnmZIM64pWFqoeNkUm2QnpOWmPwoUbN?usp=sharing

Let us know if that works for your use case (and be mindful that you need to label the missing nodes in all instances if you're using this to train a new model).

Cheers,

Talmo

**Reply by rfkova** [reply]
Wow, incredible! 

Thank you, I'll check this out tomorrow. I have a few simpler models that track different subjects well and I'm trying to tie together the output from each of them for behavioral analysis with the requested merge functionality. The skeletons do not overlap in any way and are single instance so I think this should work (I can add a connecting edge post-merge to make a single skeleton). The plan is to submit the result through SIMBA which will only take one skeleton or duplicate skeletons (last I checked).

Thank you so much for taking a crack at it! I'll let you know how it goes 😁

Best,
Ryan

**Reply by rfkova** [reply]
Hi @talmo,

Sorry, I failed to ping your handle in the above message. I'll refer you there to receive my gratitude! Essentially, I'm honored whenever you appear.

I've been messing with your script on and off all day. I'm not very fluent in python, so take this with a grain of salt and correct me if I stray from verisimilitude. 

I think your script adds an instance from `labels2 `to `labels1 `to every frame but updates the skeleton leaving nan's open to be filled. This approach makes sense when you think there might be multiple instances because as you say there are headaches therein. I have multiple "single animal" (nest, pups, dam) models each best at discerning different features of the cage, as you posit. At this stage, I would like to take the predictions generated by the model and knit together a super-creature that SIMBA can take in as a single skeleton and deduce the movement of it's parts corresponding to the dam, nest, and perhaps one day pups. 

Do you think this general approach will work to that end? 
1. I think the only thing I have to add to link the models (after you added the nodes to this skeleton) is this: `labels1.skeleton.add_edge(source = 'd_neck', destination = 'nest_c')`
2. use `labels3 = labels1.remove_user_instances()` to create an empty label object
3. ` labels3.append(dst_lf)` as you did to supply the new frame
5. fill `new_inst` with data from both `labels1 `and `labels2 `using` src_lf[0]` from each (predicted instance is always first, yes?... or is this just the way it works when you run predictions until you add more labels?)
6. ` labels3.add_instance(dst_lf, new_inst)` to drop in the new combined data

I think I can use the scaffold you made with that script and the above to cobble things together from here so I'll mark this as resolved and pop back in if I hit a wall. 

If you have any reaction to the above (caution, alarm... what have you), I would be delighted with the discourse. It does seem like iterating through a whole video knitting the skeletons on every frame takes some time... I might apply a multiprocessor approach we've implemented in some other scripts I've tinkered with... if you have any other suggestions to speed it up, lmk. 

Godspeed on your grant! 
Thank you so much for taking the time. I remain in awe of all you are doing and remain so giving to pop in to answer user questions and code. 

Best,
Ryan

**Reply by rfkova** [reply]
@talmo @roomrys,

No pressure for a quick reply, I know @roomrys is out and the grant calls 'Talmo' like the Wind Cries Mary.

So I modified the script @talmo wrote and it does do kind of what I need, but I do worry it is a little rough and it does for some reason add a second video to the labels file... which is the same as the first. This doesn't pose a problem at the moment because I can delete it in the GUI, but I would like to fix it for the future (something something labels3.remove_video... but I'm messing something up when I try that). 

I've put the script (MergeSkeletons_RK.py) [here along with my videos I'm trying to combine](https://drive.google.com/drive/folders/1bF6FJy8X8e0EWUhOVCObyZ3hd-BVN0X7?usp=drive_link). 

If anyone can take a quick look at the script, I'm sure it will not take long to identify where in my bumbling I've failed to call a class object appropriately (or at all where I ought). 

Thank you so much for your help, this is great :)

Best,
Ryan

**Reply by roomrys** [reply]
Hi @rfkova,

I made some [adjustments to your script](https://drive.google.com/file/d/1z_7W8GyA4acunJrTJLnzaM5_aMQBYvRI/view?usp=sharing).

Since you were just setting the labeled frames list to an empty list, the `Video` object was still being retained. The following removes the video (and labeled frames):

```python
  # I just need a new object to add frames too with any residuals this may leave matching the first labels file
  for video in labels3.videos:
      labels3.remove_video(video)
```

Then, when adding a new instance that came from a different `Labels` object (which also referenced a different `Video` object), another `Video` was added (even though it points to the same filename).

I also sorted your labeled frames so that when you index into the Labels.labeled_frame list we know the video(s) and frames line up (assuming you labeled the same frames in both projects):
```python
    # Sort the frames by video and frames index to ensure they're in the same order
    labels1.labeled_frames.sort(key=lambda lf: (lf.video, lf.frame_idx))
    labels2.labeled_frames.sort(key=lambda lf: (lf.video, lf.frame_idx))
```

Those were the two main changes.

Thanks,
Liezl

**Reply by rfkova** [reply]
@roomrys,

Thank you so much! I learn something every time I dip into the documentation for sleap. This was next level for me, actually getting the script written, then modified; it will be a huge help Frankenstein-ing our little critters together ^_^

Sorry for all the spam, especially while you were off.

You rock,
Ryan

**Comment by rfkova** [comment]
@roomrys,

Hi again... 

Apologies for coming in again so soon. I'm very stuck trying to script adding tracks... I just need my merged labels to all be track = '1'. I suspect (probably incorrectly) part of the problem is how we are loading a data frame without a video and adding frames to it. 

[Here's my script](https://drive.google.com/file/d/1Yn-3Or7wtGEdMpNm-ERjm6MXr9gobK2S/view?usp=sharing) to show what I'm trying. I really don't understand the documentation when it says "spawned_on: The video frame that this track was spawned on." when it suggests a string should be the input. When I add a tracked frame in the GUI I get track=Track(spawned_on=0, name='1'). If I omit the attempt to add a track and save the file, then reload it, I get predicted frames back with videos filled in for frames, but that doesn't make what I'm trying work any better, probably because my fundamental confusion with the documentation. It's definitely a 'me' problem... if you have any tips for getting to the right answer here I could use some wisdom for the future, but no pressure on that point!

Thanks for your help, as always,
Ryan

**Reply by roomrys** [reply]
Hi @rfkova,

I made some modifications to your script (v001 in this [folder](https://drive.google.com/drive/folders/1hRHNAoo5KSbw1kZ1-BOxhRttdWSbJkpE?usp=sharing)). The problem seemed to be that the `Labels.add_track` method assumed that the `Video` was already in the `LabelsDataCache` (used to speed-up frame/instance/track look-up). 
> This is the case when SLEAP calls the `Labels.add_track` method (after Labels.append(LabeledFrame)), but it isn't intuitive to do this via the API.  The API should be adding the video to the cache if it is not there already when calling `Labels.add_tracks` by itself.

The changes I made:
- use `Labels.tracks.append(Track)` instead to bypass the cache
- create one instance of `Track` outside of the `LabeledFrame`s for loop
- use `Labels.extract` to create a copy of `labels1` without any `LabeledFrame`s (or instances - the `Labels.remove_video` call was taking too long because it also looped through all the `LabeledFrame`s it needed to remove)
- use the [`tqdm`](https://tqdm.github.io/) package (which is shipped with the `sleap` development environment - available if installed from source, otherwise you will need to install this to use it) for a progress bar in lieu of the "poor man's progress bar" :)

Thanks,
Liezl

**Reply by rfkova** [reply]
@roomrys,

Thank you so much, this is so great! I'm so relieved that I wasn't 100% on the wrong track (... eyyyyyyyy...) and something the general approach (removing the video) was off-kilter. 

A few little comments in case the feedback is helpful:
1. Labels.extract is a life saver on time. However, this does reintroduce the problem that I end with two videos... However, I added a line to remove it now that I'm used to .remove_video, this was super easy <3
2. Oddly tqdm was not available in my sleap environment and I'm using 1.3.1. Knowing it was compatible, I installed it (I think with conda-forge) and now I'm no longer a poor man, for which I have you to thank!

Cheers,
Ryan

**Reply by roomrys** [reply]
The feedback is helpful! 

`tqdm` must only be in the development environment then (when installed from source) - I'll keep that in the back of my mind for next time. I've updated my claim above to be more accurate. I also found the `Labels.extract` will move the video over from the labels you are copying from, so the video removal still needs to take place afterwards (I hope I gave you the version that did the removal!).

I hope learning the API isn't too painful! Seems like you are making good progress.

Thanks,
Liezl

**Reply by rfkova** [reply]
@roomrys,

I didn't get the version with video removal but spotted it immediately and have the example from before of how to remove it! 

Not too painful. The script Talmo wrote and your feedback have made it _SO_ much easier than it would be otherwise. Definitely leaves room for excitement about it! 

I will be honest, the way the tracks treated me was kinda rough though, haha. I though I know it is partially the niche case I'm in (probably unusual to remove all videos and add stuff back). 

Thanks again for everything!
Ryan

**Reply by roomrys** [reply]
Yeah... we merged in #1407 and #1408 because it seemed pretty unintuitive to deal with `Tracks` through the API (thanks to your feedback!).

**Comment by antortjim** [comment]
Adding my two cents here :)
I slightly modified the code in [the notebok](https://colab.research.google.com/drive/1OmnmZIM64pWFqoeNkUm2QnpOWmPwoUbN?usp=sharing) to handle instances with no track (so that https://github.com/talmolab/sleap/blob/v1.3.1/sleap/io/dataset.py#L515 does not error i.e. `t` is never `None`), which is I think the same problem other people have mentioned in the last comments

```
import sleap.io.dataset

TRACK=sleap.io.dataset.Track(name="Track-0", spawned_on=0)

def merge_labels(labels1, labels2):
    """
    Combine labels stored in two different label objects

    Instances without track get assigned the track Track-0 spawned on frame 0

    Based on https://colab.research.google.com/drive/1OmnmZIM64pWFqoeNkUm2QnpOWmPwoUbN?usp=sharing#scrollTo=52ojKlgcz6ns
    """
    
    nodes1_inds = np.array([labels1.skeleton.node_names.index(node) for node in labels2.skeleton.node_names])
    
    for src_lf in labels2:
        if (src_lf.video, src_lf.frame_idx) in labels1:
            # If the video/frame_idx is labeled in the base dataset, we'll just add to it
            dst_lf = labels1[(src_lf.video, src_lf.frame_idx)]

        else:
            # Create a new labeled frame and add it to the base dataset
            dst_lf = sleap.LabeledFrame(video=src_lf.video, frame_idx=src_lf.frame_idx)
            labels1.append(dst_lf)

        for inst in src_lf:

            if inst.track is None:
                track = TRACK
            else:
                track = inst.track
        
            # Create new instance with unified skeleton
            if type(inst) == sleap.PredictedInstance:
                old_pts = inst.numpy()
                old_scores = inst.scores

                new_pts = np.full((len(labels1.skeleton), 2), np.nan)
                new_pts[nodes1_inds] = old_pts
                new_scores = np.full((len(labels1.skeleton)), np.nan)
                new_scores[nodes1_inds] = old_scores

                new_inst = sleap.PredictedInstance.from_numpy(new_pts, new_scores, instance_score=inst.score, skeleton=labels1.skeleton, track=track)

            elif type(inst) == sleap.Instance:
                old_pts = inst.numpy()

                new_pts = np.full((len(labels1.skeleton), 2), np.nan)
                new_pts[nodes1_inds] = old_pts

                new_inst = sleap.Instance.from_numpy(new_pts, skeleton=labels1.skeleton, track=track)

            # TODO: Handle 100% overlapping instances to prevent trivial duplicates

            # Add the new instance to the destination frame in the base labels
            labels1.add_instance(dst_lf, new_inst)

            if new_inst.track not in labels1.tracks:
                # TODO: Rename tracks with the duplicate names to avoid confusion
                labels1.tracks.append(new_inst.track)


    return labels1
```
